src/igmpproxy.c

```diff
@@ -140,16 +143,42 @@ int main( int ArgCn, char *ArgVc[] ) {
             break;
         }
 
+        // Open /dev/null before chrooting.
+        if (!NotAsDaemon) {
+            devnull = open(_PATH_DEVNULL, O_RDWR, 0);
+            if (devnull == -1)
+                my_log(LOG_ERR, 0, "unable to open %s", _PATH_DEVNULL);
+        }
+
+        config = getCommonConfig();
+
+        if (strlen(config->user) != 0) {
+            pw = getpwnam(config->user);
+            if (pw == NULL)
+                my_log(LOG_ERR, 0, "unknown user %s", config->user);
+        }
+
+        if (strlen(config->chroot) != 0)
+            if (chroot(config->chroot) != 0 || chdir("/") != 0)
+                my_log(LOG_ERR, 0, "unable to chroot to %s",
+                  config->chroot);
+
+        if (strlen(config->user) != 0)
+            if (setgroups(1, &pw->pw_gid) != 0 ||
+              setresgid(pw->pw_gid, pw->pw_gid, pw->pw_gid) != 0 ||
+              setresuid(pw->pw_uid, pw->pw_uid, pw->pw_uid) != 0)
+                my_log(LOG_ERR, 0, "unable to drop privileges to %s",
+                  config->user);
+
         if ( !NotAsDaemon ) {
 
             // Only daemon goes past this line...
             if (fork()) exit(0);
 
             // Detach daemon from terminal
             if ( close( 0 ) < 0 || close( 1 ) < 0 || close( 2 ) < 0
-                || open( "/dev/null", 0 ) != 0 || dup2( 0, 1 ) < 0 || dup2( 0, 2 ) < 0
-                || setpgid( 0, 0 ) < 0
-            ) {
+                || dup2( devnull, 0 ) < 0 || dup2( devnull, 1 ) < 0
+                || dup2( devnull, 2 ) < 0 || setpgid( 0, 0 ) < 0 ) {
```

Comment:
This your change introduces a new resource leak. `devnull` file descriptor is not used after these calls and stay open forever when running in daemon mode.

Reply:
Good catch! Issue has been addressed.